src/meta/processors/admin/Balancer.cpp

```diff
@@ -675,5 +794,84 @@ void Balancer::simplifyLeaderBalnacePlan(GraphSpaceID spaceId, LeaderBalancePlan
     }
 }
 
+bool Balancer::fetchHostLoadingBySpace(GraphSpaceID spaceId, int32_t replicaFactor) {
+    std::unique_ptr<kvstore::KVIterator> iter;
+    const auto& prefix = MetaServiceUtils::partPrefix(spaceId);
+    auto code = kv_->prefix(kDefaultSpaceId, kDefaultPartId, prefix, &iter);
+    if (code != kvstore::ResultCode::SUCCEEDED) {
+        LOG(ERROR) << "List Parts Failed";
+        return false;
+    }
+
+    std::unordered_map<HostAddr, int32_t> hostLoading;
+    while (iter->valid()) {
+        auto hosts = MetaServiceUtils::parsePartVal(iter->val());
+        for (auto& host : hosts) {
+            hostLoading[host]++;
+        }
+        iter->next();
+    }
+
+    for (auto it = hostLoading.begin(); it != hostLoading.end(); it++) {
+        int32_t min = it->second / replicaFactor;
+        auto pair = Bounds(min, min + 1);
+        LOG(INFO) << "Host: " << it->first << " Bounds: " << min << " : " << min + 1;
+        hostBounds_[it->first] = std::move(pair);
+    }
+    return true;
+}
+
+bool Balancer::calculationBounds(GraphSpaceID spaceId, int32_t replicaFactor) {
+    std::unique_ptr<kvstore::KVIterator> iter;
+    const auto& prefix = MetaServiceUtils::partPrefix(spaceId);
+    auto code = kv_->prefix(kDefaultSpaceId, kDefaultPartId, prefix, &iter);
+    if (code != kvstore::ResultCode::SUCCEEDED) {
+        LOG(ERROR) << "List Parts Failed";
+        return false;
+    }
+
+    std::unordered_map<HostAddr, int32_t> hostLoading;
+    PartAllocation partAddresses;
+    while (iter->valid()) {
+        auto partId = MetaServiceUtils::parsePartKeyPartId(iter->key());
+        auto hosts = MetaServiceUtils::parsePartVal(iter->val());
+        for (auto host : hosts) {
+            hostLoading[host]++;
+            partAddresses[partId].emplace_back(host);
+        }
+        iter->next();
+    }
+
+    std::unordered_map<PartitionID, std::vector<int32_t>> partLoading;
+    for (auto partIter = partAddresses.begin(); partIter != partAddresses.end(); partIter++) {
+        auto part = partIter->first;
+        for (auto address : partIter->second) {
+            auto hostIter = hostLoading.find(address);
+            if (hostIter == hostLoading.end()) {
+                LOG(ERROR) << "Can't find host " << address;
+                return false;
+            }
+
+            VLOG(3) << "Part: " << part << " size: " << hostIter->second;
+            partLoading[part].emplace_back(hostIter->second);
+        }
+    }
```

Comment:
Don't do disk io and calculate for each time one partition is transferred.

Reply:
really good idea